client/client.go

```diff
@@ -376,6 +358,68 @@ func (c *client) Close() {
 	}
 }
 
+func (c *client) setServiceMode(newMode pdpb.ServiceMode) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if newMode == c.serviceMode {
+		return
+	}
+
+	log.Info("changing service mode", zap.String("old-mode", pdpb.ServiceMode_name[int32(c.serviceMode)]),
+		zap.String("new-mode", pdpb.ServiceMode_name[int32(newMode)]))
+
+	if newMode == pdpb.ServiceMode_UNKNOWN_SVC_MODE {
+		log.Warn("intend to switch to unknown service mode. do nothing")
+		return
+	}
+
+	var newTSOCli *tsoClient
+	tsoSvcDiscovery := c.tsoSvcDiscovery
+	ctx, cancel := context.WithCancel(c.ctx)
+	wg := &sync.WaitGroup{}
```

Comment:
Why pass the `wg` as a parameter rather than creating the wait group inside each constructor function?

Reply:
It is shared acroos tsoClient and tsoServiceDiscovery. Yes, I can separate them.

**Separted**.